library/common/extensions/filters/http/local_error/filter.cc

```diff
@@ -15,25 +17,75 @@ namespace Extensions {
+  switch (status) {
+  case 408:
+    headers.addCopy(Http::InternalHeaders::get().ErrorCode, std::to_string(ENVOY_REQUEST_TIMEOUT));
+    break;
+  case 413:
+    headers.addCopy(Http::InternalHeaders::get().ErrorCode,
+                    std::to_string(ENVOY_BUFFER_LIMIT_EXCEEDED));
+    break;
+  case 503:
+    headers.addCopy(Http::InternalHeaders::get().ErrorCode,
+                    std::to_string(ENVOY_CONNECTION_FAILURE));
+    break;
+  default:
+    headers.addCopy(Http::InternalHeaders::get().ErrorCode, std::to_string(ENVOY_UNDEFINED_ERROR));
+  }
+
+  headers.setStatus(218);
+}
+
+// Envoy Mobile surfaces non-"OK" local responses as errors via callbacks rather than an HTTP
+// response. Here that response's "real" status code is mapped to an Envoy Mobile error code
+// which is passed through the response chain as a header. The status code is updated with
+// the sentinel value, 218 ("This is fine").
+void mapLocalGrpcResponseToError(Grpc::Status::GrpcStatus status,
+                                 Http::ResponseHeaderMap& headers) {
+  return mapLocalHttpResponseToError(Grpc::Utility::grpcToHttpStatus(status), headers);
+}
+
+} // namespace
+
 LocalErrorFilter::LocalErrorFilter() {}
 
 Http::FilterHeadersStatus LocalErrorFilter::encodeHeaders(Http::ResponseHeaderMap& headers,
                                                           bool end_stream) {
-  uint64_t response_status = Http::Utility::getResponseStatus(headers);
-  bool success = Http::CodeUtility::is2xx(response_status);
+  // HACK: currently Envoy sends local replies in cases where an error ought to be
+  // surfaced via the error path. See https://github.com/lyft/envoy-mobile/issues/460
 
-  // TODO: ***HACK*** currently Envoy sends local replies in cases where an error ought to be
-  // surfaced via the error path. There are ways we can clean up Envoy's local reply path to
-  // make this possible, but nothing expedient. For the immediate term this is our only real
-  // option. See https://github.com/lyft/envoy-mobile/issues/460
+  // Absence of EnvoyUpstreamServiceTime header implies this is a local reply, which we potentially
+  // map to a stream error.
+  if (!headers.get(Http::Headers::get().EnvoyUpstreamServiceTime).empty()) {
+    return Http::FilterHeadersStatus::Continue;
+  }
 
-  // Absence of EnvoyUpstreamServiceTime header implies this is a local reply, which we treat as
-  // a stream error.
-  if (!success && headers.get(Http::Headers::get().EnvoyUpstreamServiceTime).empty()) {
+  const auto grpc_status = Grpc::Common::getGrpcStatus(headers);
+  if (grpc_status && grpc_status.value() != Grpc::Status::WellKnownGrpcStatus::Ok) {
+    ASSERT(end_stream,
+           "Local gRPC responses must consist of a single headers frame. If Envoy changes "
+           "this expectation, this code needs to be updated.");
+    ENVOY_LOG(debug, "intercepted local GRPC response");
+    grpcError_ = true;
```

Comment:
It looks like this is never read, is it needed?

Reply:
Actually you're right, it's not; good catch. I was using it previously, but was able to simplify things. I'll go ahead and remove.